### PR TITLE
[SPARK-31029] Avoid using global execution context in driver main thread for YarnSchedulerBackend

### PR DESCRIPTION
**Pull request by shanyu**
#31029 # What changes were proposed in this pull request?
In YarnSchedulerBackend, we should avoid using the global execution context for its Future. Otherwise if user's Spark application also uses global execution context for its Future, the user is facing indeterministic behavior in terms of the thread's context class loader.

### Why are the changes needed?
When running tpc-ds test (https://github.com/databricks/spark-sql-perf), occasionally we see error related to class not found:

2020-02-04 20:00:26,673 ERROR yarn.ApplicationMaster: User class threw exception: scala.ScalaReflectionException: class com.databricks.spark.sql.perf.ExperimentRun in JavaMirror with
sun.misc.Launcher$AppClassLoader@28ba21f3 of type class sun.misc.Launcher$AppClassLoader with classpath [...]
and parent being sun.misc.Launcher$ExtClassLoader@3ff5d147 of type class sun.misc.Launcher$ExtClassLoader with classpath [...]
and parent being primordial classloader with boot classpath [...] not found.

This is the root cause for the problem:

Spark driver starts ApplicationMaster in the main thread, which starts a user thread and set MutableURLClassLoader to that thread's ContextClassLoader.
  userClassThread = startUserApplication()

The main thread then setup YarnSchedulerBackend RPC endpoints, which handles these calls using scala Future with the default global ExecutionContext:
  doRequestTotalExecutors
  doKillExecutors

So for the main thread and user thread, whoever starts the future first get a chance to set ContextClassLoader to the default thread pool:

- If main thread starts a future to handle doKillExecutors() before user thread does then the default thread pool thread's ContextClassLoader would be the default (AppClassLoader).
- If user thread starts a future first then the thread pool thread will have MutableURLClassLoader.

Note that only MutableURLClassLoader can load user provided class for the Spark app, you will see errors related to class not found if the ContextClassLoader is AppClassLoader.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Existing unit tests and manual tests

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @Ngone51 @tgravescs

**Comment by tgravescs** [comment]
I'll look in more detail, looks like the implicit is on the execution context for the Future callback.  @shanyu  can you update the description with more details on the exact fix  and I think we should put a comment there at least so people looking at this in the future know without having to search.

**Comment by tgravescs** [comment]
ok to test

**Comment by Ngone51** [comment]
Just out of curiosity, why Jenkins test job success within 20 mins?

**Comment by tgravescs** [comment]
probably because it only ran the yarn tests

**Comment by shanyu** [comment]
Hi @tgravescs are we good to merge this PR?

**Comment by tgravescs** [comment]
I didn't finish reviewing.  Please update the description to describe the fix and add a comment by the line schedulerEndpointEC definition.

File: resource-managers/yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala [hunk 1]
```diff
@@ -66,6 +65,9 @@ private[spark] abstract class YarnSchedulerBackend(
     YarnSchedulerBackend.ENDPOINT_NAME, yarnSchedulerEndpoint)
 
   private implicit val askTimeout = RpcUtils.askRpcTimeout(sc.conf)
+  private implicit val schedulerEndpointEC =
+    ExecutionContext.fromExecutorService(
```
Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala:69]: nit put this line on previous line

**Comment by shanyu** [comment]
@tgravescs Thanks for reviewing! I've updated the description and added a comment to the implicit execution context.

File: resource-managers/yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala [hunk 2]
```diff
@@ -67,6 +66,13 @@ private[spark] abstract class YarnSchedulerBackend(
 
   private implicit val askTimeout = RpcUtils.askRpcTimeout(sc.conf)
 
+  /**
+   * Declare implicit single thread execution context for futures doRequestTotalExecutors and
+   * doKillExecutors below, avoiding using the global execution context that may cause conflict
+   * with user code's execution of futures */
```
Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala:72]: nit put the last */ on the next line.

Comment (shanyu) [reply resource-managers/yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala]: sorry my bad, fixed.

**Review by tgravescs — APPROVED** [review]

**Comment by tgravescs** [comment]
pushing to master, thanks @shanyu

**Pull request closed without merging**